pkg/auth/dependency/interaction/provider.go

```diff
@@ -0,0 +1,95 @@
+package interaction
+
+import (
+	gotime "time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/skygeario/skygear-server/pkg/core/authn"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/time"
+)
+
+type IdentityProvider interface {
+	Get(userID string, typ authn.IdentityType, id string) (*IdentityInfo, error)
+	GetByClaims(typ authn.IdentityType, claims map[string]interface{}) (string, *IdentityInfo, error)
+}
+
+type AuthenticatorProvider interface {
+	Get(userID string, typ AuthenticatorType, id string) (*AuthenticatorInfo, error)
+	List(userID string, typ AuthenticatorType) ([]*AuthenticatorInfo, error)
+	Authenticate(userID string, spec AuthenticatorSpec, state *map[string]string, secret string) (*AuthenticatorInfo, error)
+}
+
+// TODO(interaction): configurable lifetime
+const interactionIdleTimeout = 5 * gotime.Minute
+
+type Provider struct {
+	Store         Store
+	Time          time.Provider
+	Logger        *logrus.Entry
+	Identity      IdentityProvider
+	Authenticator AuthenticatorProvider
+	Config        *config.AuthenticationConfiguration
+}
+
+func (p *Provider) GetInteraction(token string) (*Interaction, error) {
+	i, err := p.Store.Get(token)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
Why do we need to fetch Identity/PrimaryAuthenticator/SecondaryAuthenticator?

Reply:
We should be using the most up-to-date information of authenticator, in case of removal, password changes, etc.